tools/aiecc/aiecc/main.py

```diff
@@ -120,31 +120,42 @@ def process_core(core):
             do_call(['clang', '-O2', '--target=aie', file_core_obj, me_basic_o, libm,
             '-Wl,-T,'+file_core_ldscript, '-o', file_core_elf])
 
-    # Compile each core in parallel
-    # Parallel(n_jobs=8, require='sharedmem')(delayed(process_core)(core) for core in cores)
-    for core in cores:
-      process_core(core)
 
-    # Generate the included host interface
-    file_physical = os.path.join(tmpdirname, 'input_physical.mlir')
-    if(opts.pathfinder):
-      do_call(['aie-opt', '--aie-create-pathfinder-flows', file_with_addresses, '-o', file_physical]);
+    def process_arm_cgen():
+      # Generate the included host interface
+      file_physical = os.path.join(tmpdirname, 'input_physical.mlir')
+      if(opts.pathfinder):
+        do_call(['aie-opt', '--aie-create-pathfinder-flows', file_with_addresses, '-o', file_physical]);
+      else:
+        do_call(['aie-opt', '--aie-create-flows', file_with_addresses, '-o', file_physical]);
+      file_inc_cpp = os.path.join(tmpdirname, 'aie_inc.cpp')
+      do_call(['aie-translate', '--aie-generate-xaie', file_physical, '-o', file_inc_cpp])
+
+      # Lastly, compile the generated host interface with any ARM code.
+      cmd = ['clang','--target=aarch64-linux-gnu', '-std=c++11']
+      if(opts.sysroot):
+        cmd += ['--sysroot=%s' % opts.sysroot]
+      cmd += ['-I%s/opt/xaiengine/include' % opts.sysroot]
+      cmd += ['-L%s/opt/xaiengine/lib' % opts.sysroot]
+      cmd += ['-I%s' % tmpdirname]
+      cmd += ['-fuse-ld=lld','-rdynamic','-lxaiengine','-lmetal','-lopen_amp','-ldl']
+
+      if(len(opts.arm_args) > 0):
+        do_call(cmd + opts.arm_args)
+
+
+    if (opts.nthreads == True):
+      with ThreadPool() as thdpool:
+          thdpool.map(process_core, (cores))
+          thdpool.apply_async(process_arm_cgen)
```

Comment:
How does this code prioritize the arm_cgen task, which we expect to be long?

Reply:
I cannot test now, but if I remember correctly, putting the apply_async() before the map() has almost the same performance. But you are right, spawning the arm_code_gen() in advance should be theoretically faster. Will change that.

Reply:
It could be worth a comment like "prefer to run process_arm_cgen first, since it typically takes longer than process_core".